CppImport/src/macro/MetaDefinitionManager.cpp

```diff
@@ -0,0 +1,249 @@
+		Q_ASSERT(namespaceModule); //if (!namespaceModule) return root_;
+
+		// try to find the module (includes macro containers) to put this macro in
+		result = StaticStuff::findDeclaration(namespaceModule->modules(), fileName);
+
+		// if no module could be found; try to find an appropriate class to put this macro in
+		if (!result) result = StaticStuff::findDeclaration(namespaceModule->classes(), fileName);
+
+		// if no existing place could be found: create a new module (macro container) and put the macro in there
+		if (!result)
+		{
+			result = new OOModel::Module(fileName);
+			namespaceModule->modules()->append(result);
+		}
+	}
+	else
+	{
+		result = StaticStuff::findDeclaration(root_->modules(), "notenvision");
+
+		if (!result)
+		{
+			result = new OOModel::Module("notenvision");
+			root_->modules()->append(result);
+		}
+	}
+
+	Q_ASSERT(result);
+	return result;
+}
+
+OOModel::MetaDefinition* MetaDefinitionManager::metaDefinition(const clang::MacroDirective* md)
+{
+	QString h = definitionManager_->hash(md);
+
+	if (!metaDefinitions_.contains(h))
+		return nullptr;
+
+	return metaDefinitions_.value(h);
+}
+
+void MetaDefinitionManager::createMetaDef(QVector<Model::Node*> nodes, MacroExpansion* expansion, NodeMapping* mapping,
+														QVector<MacroArgumentInfo>& arguments)
+{
+	if (!metaDefinition(expansion->definition))
+	{
+		auto metaDef = new OOModel::MetaDefinition(definitionManager_->definitionName(expansion->definition));
+		metaDefinitions_.insert(definitionManager_->hash(expansion->definition), metaDef);
+
+		// add formal arguments based on the expansion definition
+		for (auto argName : clang_->argumentNames(expansion->definition))
+			metaDef->arguments()->append(new OOModel::FormalMetaArgument(argName));
+
+		auto metaDefParent = metaDefinitionParent(expansion->definition);
+
+		// check whether this expansion is not a potential partial begin macro specialization
+		if (auto beginChild = xMacroManager_->partialBeginChild(expansion))
+			xMacroManager_->handlePartialBeginSpecialization(metaDefParent, metaDef, expansion, beginChild);
+		else
+		{
+			// case: meta definition is not a partial begin macro specialization
+
+			if (nodes.size() > 0)
+			{
+				// create a new context with type equal to the first node's context
+				auto actualContext = StaticStuff::actualContext(mapping->original(nodes.first()));
+				metaDef->setContext(StaticStuff::createContext(actualContext));
+
+				// clone and add nodes to the metaDef
+				for (auto n : nodes)
+				{
+					NodeMapping childMapping;
+					auto cloned = StaticStuff::cloneWithMapping(mapping->original(n), &childMapping);
+
+					lexicalHelper_->applyLexicalTransformations(cloned, &childMapping,
+																					 clang_->argumentNames(expansion->definition));
+
+					insertChildMetaCalls(expansion, &childMapping);
+
+					if (removeUnownedNodes(cloned, expansion, &childMapping))
+						continue;
+
```

Comment:
You leave an empty line after every statement here. This defeats the purpose of empty lines to group sections of code. In general there are too many empty lines in many parts of the code. Please go over all the code, and reduce (not elimintate) the number of empty lines, so that they serve their true purpose.